### ISSUE TITLE
__PHP_Incomplete_Class + Segfault

### ISSUE THREAD

**Issue by SirSnyder** [krakjoe/pthreads#666]
Hi Joe,

my application loads some classes at first. With a call of Thread::start(), many warnings like "Warning: Class __PHP_Incomplete_Class has no unserializer in Unknown on line 0" appear in the console.

[coredump.txt](https://github.com/krakjoe/pthreads/files/724105/coredump.txt)

Example code will follow.

Ubuntu with php 7.1.2-dev and pthreads latest master. 

 Thx

**Comment by SirSnyder** [comment]
Segfault example code

```
function objectHash($object)
    {
        return crc32(spl_object_hash($object));
    }

    trait ClassInfo
    {
        /** @var int */
        private $intObjectHash;
    }

    class ThreadedMap extends Threaded
    {
        use ClassInfo;
    }

    class LiquidObject extends Threaded
    {
        use ClassInfo;
    }

    class ManagedInstanceProvider extends Threaded
    {
        use ClassInfo;

        const FREE = 0;
        const USED = 1;

        /** @var \Threaded */
        private $objIndex;

        /** @var \Threaded */
        private $objStack;

        /** @var \Threaded */
        private $objHashMap;

        /** @var bool */
        private $blnActive;

        /** @var string */
        private $strClassname;

        /** @var int */
        private $intLastIndex;

        /** @var int */
        private $intNumFreeIndizes;


        public function __construct(?string $classname)
        {
            $this->strClassname         = $classname;
            $this->intLastIndex         = 0;
            $this->intNumFreeIndizes    = 0;
            $this->blnActive            = false;
        }


        public function manage():void
        {
            $this->objStack->synchronized(function ()
            {
                if($this->intNumFreeIndizes === 0 && $this->objStack->count() === $this->intLastIndex)
                {
                    $objInstance = $this->createInstance();

                    $this->objStack[] = $objInstance;

                    $intPosition = $this->objStack->count() - 1;

                    $this->objIndex[$intPosition] = self::FREE;

                    $this->objHashMap[objectHash($objInstance)] = $intPosition;

                    $this->objStack->notify();
                }
            });
        }


        public function get()
        {
            return $this->objStack->synchronized(function ()
            {
                if(!$this->isActive())
                {
                    $this->activate();
                }
                $objInstance = null;

                while($objInstance === null)
                {
                    if($this->intNumFreeIndizes > 0)
                    {
                        $objInstance = $this->findFreeInstance();
                        continue;
                    }

                    if($this->objStack->count() > $this->intLastIndex)
                    {
                        // get instance
                        $objInstance = $this->objStack[$this->intLastIndex];

                        // mark instance used
                        $this->objIndex[$this->intLastIndex] = self::USED;

                        var_dump('get instance');

                        ++$this->intLastIndex;
                        continue;
                    }
                    $this->objStack->wait(0);
                }
                return $objInstance;
            });
        }


        public function isActive():bool
        {
            return $this->blnActive;
        }


        public function setActive(bool $active):void
        {
            $this->blnActive = $active;
        }


        public function setClassname(string $classname):void
        {
            $this->strClassname = $classname;
        }


        public function activate():void
        {
            $this->synchronized(function ()
            {
                if($this->isActive())
                {
                    return;
                }
                $this->objIndex     = new \Threaded();
                $this->objStack     = new ThreadedMap();// \Threaded();
                $this->objHashMap   = new \Threaded();

                $this->setActive(true);
            });
        }


        private function findFreeInstance()
        {
            foreach($this->objIndex as $position => $value)
            {
                if($value === self::FREE)
                {
                    $this->objIndex[$position] = self::USED;

                    --$this->intNumFreeIndizes;

                    return $this->objStack[$position];
                }
            }
            return null;
        }


        private function createInstance()
        {
            $objReflectionClass = new ReflectionClass($this->strClassname);

            $blnCallWithConstructor = false;

            /** @var \ReflectionMethod $objConstructor */
            $objConstructor = $objReflectionClass->getConstructor();

            if($objConstructor === null || ($objConstructor->getNumberOfRequiredParameters() === 0 && !$objConstructor->isPrivate()))
            {
                $blnCallWithConstructor = true;
            }
            $objInstance = null;

            if($blnCallWithConstructor)
            {
                $objInstance = $objReflectionClass->newInstance();
            }
            else
            {
                $objInstance = $objReflectionClass->newInstanceWithoutConstructor();
            }
            return $objInstance;
        }
    }

    class ManagedInstanceImmatureProviderHolder extends Threaded
    {
        use ClassInfo;

        /** @var int */
        private $intPosition;


        public function __construct()
        {
            $this->intPosition  = 0;
        }


        public function numFreeProviders():int
        {
            return $this->count() - 2 - $this->intPosition;
        }


        public function getProvider():ManagedInstanceProvider
        {
            return $this->synchronized(function ()
            {
                while($this->numFreeProviders() === 0)
                {
                    $this->wait(0);
                }
                return $this[$this->intPosition++];
            });
        }


        public function addProvider(ManagedInstanceProvider $provider)
        {
            $this->synchronized(function () use ($provider)
            {
                $this[] = $provider;
                $this->notify();
            });
        }
    }

    class ManagedInstanceProviderDataShare extends Threaded
    {
        use ClassInfo;

        /** @var \Threaded */
        private $objProviderHolder;

        /** @var ManagedInstanceImmatureProviderHolder */
        private $objImmatureProviderHolder;


        public function __construct()
        {
            $this->objProviderHolder            = null;
            $this->objImmatureProviderHolder    = null;
        }

        public function getProviderHolder():\Threaded
        {
            if($this->objProviderHolder === null)
            {
                $this->objProviderHolder = new \Threaded();
            }
            return $this->objProviderHolder;
        }


        public function getImmatureProviderHolder():ManagedInstanceImmatureProviderHolder
        {
            if($this->objImmatureProviderHolder === null)
            {
                $this->objImmatureProviderHolder = new ManagedInstanceImmatureProviderHolder();
            }
            return $this->objImmatureProviderHolder;
        }
    }


    class ManagedInstanceProviderManager extends Threaded
    {
        use ClassInfo;

        /** @var ManagedInstanceProviderDataShare */
        private static $objDataShare;


        public static function globalSharedContainer($container)
        {
            self::$objDataShare = $container;
        }


        protected function getProviderHolder():\Threaded
        {
            return self::$objDataShare->getProviderHolder();
        }


        protected function getImmatureProviderHolder():ManagedInstanceImmatureProviderHolder
        {
            return self::$objDataShare->getImmatureProviderHolder();
        }


        public function getInstanceProviderFor(string $classname):ManagedInstanceProvider
        {
            $objProviderHolder = self::getProviderHolder();

            /** @var ManagedInstanceProvider $objManagedInstanceProvider */
            $objManagedInstanceProvider = $objProviderHolder->synchronized(function() use ($classname, $objProviderHolder)
            {
                if(!isset($objProviderHolder[$classname]))
                {
                    /** @var ManagedInstanceImmatureProviderHolder $objProviderHolder */
                    $objImmatureProviderHolder = $this->getImmatureProviderHolder();

                    /** @var ManagedInstanceProvider $objManagedInstanceProvider */
                    $objManagedInstanceProvider = $objImmatureProviderHolder->getProvider();

                    $objManagedInstanceProvider->setClassname($classname);
                    $objManagedInstanceProvider->activate();

                    $objProviderHolder[$classname] = $objManagedInstanceProvider;

                    return $objManagedInstanceProvider;
                }
                return $objProviderHolder[$classname];
            });
            return $objManagedInstanceProvider;
        }


        public function generateProviders()
        {
            /** @var ManagedInstanceImmatureProviderHolder $objProviderHolder */
            $objProviderHolder = $this->getImmatureProviderHolder();

            while($objProviderHolder->numFreeProviders() < 10)
            {
                /** @var ManagedInstanceProvider $objProvider */
                $objProvider = (new ReflectionClass(ManagedInstanceProvider::class))->newInstance(null);

                $objProviderHolder->addProvider($objProvider);
            }
        }


        public function supplyInstancesToProviders():void
        {
            /** @var ManagedInstanceProvider $objManagedInstanceProvider */
            foreach(self::getProviderHolder() as $objManagedInstanceProvider)
            {
                $objManagedInstanceProvider->manage();
            }
        }
    }

    class SystemLoader
    {
        public static $dataShare;
    }

    class Runtime
    {
        use ClassInfo;

        /** @var ManagedInstanceProviderManager */
        public static $manager;
    }
    SystemLoader::$dataShare = new ManagedInstanceProviderDataShare();
    Runtime::$manager = new ManagedInstanceProviderManager();
    Runtime::$manager->globalSharedContainer(SystemLoader::$dataShare);
    Runtime::$manager->generateProviders();

    class TestThread extends Thread
    {
        use ClassInfo;

        public function run()
        {
            $objInstanceProvider = Runtime::$manager->getInstanceProviderFor(LiquidObject::class);

            $counter = 0;

            while($counter++ < 256)
            {
                $objInstanceProvider->get();
            }
        }
    }
    $objTestThread = new TestThread();
    $objTestThread->start();

    $objMutex = new Threaded();
    $objMutex->synchronized(function () use ($objMutex)
    {
        while(true)
        {
            Runtime::$manager->generateProviders();
            Runtime::$manager->supplyInstancesToProviders();

            $objMutex->wait(100000);
        }
    });
```

gdb backtrace

```
#0  0x00007f5c010d1337 in kill () at ../sysdeps/unix/syscall-template.S:81
#1  0x0000000000af6d91 in zend_mm_panic (message=0x11f1a15 "zend_mm_heap corrupted") at /home/vagrant/php-src/Zend/zend_alloc.c:361
#2  0x0000000000af8bde in zend_mm_free_heap (heap=0x7f5bfe600040, ptr=0x7f5bfd263080, __zend_filename=0x11f9988 "/home/vagrant/php-src/Zend/zend_hash.c", __zend_lineno=867,
    __zend_orig_filename=0x0, __zend_orig_lineno=0) at /home/vagrant/php-src/Zend/zend_alloc.c:1374
#3  0x0000000000afb5fa in _efree (ptr=0x7f5bfd263080, __zend_filename=0x11f9988 "/home/vagrant/php-src/Zend/zend_hash.c", __zend_lineno=867, __zend_orig_filename=0x0,
    __zend_orig_lineno=0) at /home/vagrant/php-src/Zend/zend_alloc.c:2433
#4  0x0000000000b525e0 in zend_hash_do_resize (ht=0x7f5bfd286ae0) at /home/vagrant/php-src/Zend/zend_hash.c:867
#5  0x0000000000b520f5 in _zend_hash_index_add_or_update_i (ht=0x7f5bfd286ae0, h=8, pData=0x7f5bfe6145e0, flag=1,
    __zend_filename=0x7f5bfdefe920 "/home/vagrant/src/pthreads/src/store.c", __zend_lineno=358) at /home/vagrant/php-src/Zend/zend_hash.c:796
#6  0x0000000000b52395 in _zend_hash_index_update (ht=0x7f5bfd286ae0, h=8, pData=0x7f5bfe6145e0, __zend_filename=0x7f5bfdefe920 "/home/vagrant/src/pthreads/src/store.c",
    __zend_lineno=358) at /home/vagrant/php-src/Zend/zend_hash.c:836
#7  0x00007f5bfdef5ae9 in pthreads_store_write (object=0x7f5bfe614680, key=0x0, write=0x7f5bfe6145e0) at /home/vagrant/src/pthreads/src/store.c:358
#8  0x00007f5bfdef845d in pthreads_write_property (object=0x7f5bfe614680, member=0x0, value=0x7f5bfe6145e0, cache=0x0) at /home/vagrant/src/pthreads/src/handlers.c:232
#9  0x00007f5bfdef8504 in pthreads_write_dimension (object=0x7f5bfe614680, member=0x0, value=0x7f5bfe6145e0) at /home/vagrant/src/pthreads/src/handlers.c:246
#10 0x0000000000ba0064 in zend_assign_to_object_dim (object=0x7f5bfe614680, dim=0x0, value=0x7f5bfe6145e0) at /home/vagrant/php-src/Zend/zend_execute.c:1097
#11 0x0000000000bdb3b9 in ZEND_ASSIGN_DIM_SPEC_VAR_UNUSED_OP_DATA_CV_HANDLER () at /home/vagrant/php-src/Zend/zend_vm_execute.h:21634
#12 0x0000000000ba6246 in execute_ex (ex=0x7f5bfe614590) at /home/vagrant/php-src/Zend/zend_vm_execute.h:429
#13 0x00007f5bfdee7a26 in pthreads_execute_ex (data=0x7f5bfe614590) at /home/vagrant/src/pthreads/php_pthreads.c:146
#14 0x0000000000b21904 in zend_call_function (fci=0x7ffd6cc3b1a0, fci_cache=0x7ffd6cc3b1d8) at /home/vagrant/php-src/Zend/zend_execute_API.c:828
#15 0x00007f5bfdeeac8a in zim_Threaded_synchronized (execute_data=0x7f5bfe614530, return_value=0x7ffd6cc3b230) at ./classes/threaded.h:206
#16 0x0000000000ba807c in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER () at /home/vagrant/php-src/Zend/zend_vm_execute.h:970
#17 0x0000000000ba6246 in execute_ex (ex=0x7f5bfe6144b0) at /home/vagrant/php-src/Zend/zend_vm_execute.h:429
#18 0x00007f5bfdee7a26 in pthreads_execute_ex (data=0x7f5bfe6144b0) at /home/vagrant/src/pthreads/php_pthreads.c:146
#19 0x0000000000ba7fa3 in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER () at /home/vagrant/php-src/Zend/zend_vm_execute.h:949
#20 0x0000000000ba6246 in execute_ex (ex=0x7f5bfe614420) at /home/vagrant/php-src/Zend/zend_vm_execute.h:429
#21 0x00007f5bfdee7a26 in pthreads_execute_ex (data=0x7f5bfe614420) at /home/vagrant/src/pthreads/php_pthreads.c:146
#22 0x0000000000ba7fa3 in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER () at /home/vagrant/php-src/Zend/zend_vm_execute.h:949
#23 0x0000000000ba6246 in execute_ex (ex=0x7f5bfe614370) at /home/vagrant/php-src/Zend/zend_vm_execute.h:429
#24 0x00007f5bfdee7a26 in pthreads_execute_ex (data=0x7f5bfe614370) at /home/vagrant/src/pthreads/php_pthreads.c:146
#25 0x0000000000b21904 in zend_call_function (fci=0x7ffd6cc3b620, fci_cache=0x7ffd6cc3b658) at /home/vagrant/php-src/Zend/zend_execute_API.c:828
#26 0x00007f5bfdeeac8a in zim_Threaded_synchronized (execute_data=0x7f5bfe614310, return_value=0x7ffd6cc3b6b0) at ./classes/threaded.h:206
#27 0x0000000000ba807c in ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER () at /home/vagrant/php-src/Zend/zend_vm_execute.h:970
#28 0x0000000000ba6246 in execute_ex (ex=0x7f5bfe614030) at /home/vagrant/php-src/Zend/zend_vm_execute.h:429
#29 0x00007f5bfdee7a26 in pthreads_execute_ex (data=0x7f5bfe614030) at /home/vagrant/src/pthreads/php_pthreads.c:146
#30 0x0000000000ba644b in zend_execute (op_array=0x7f5bfe692000, return_value=0x0) at /home/vagrant/php-src/Zend/zend_vm_execute.h:474
#31 0x0000000000b3cda6 in zend_execute_scripts (type=8, retval=0x0, file_count=3) at /home/vagrant/php-src/Zend/zend.c:1475
#32 0x0000000000a7fbb3 in php_execute_script (primary_file=0x7ffd6cc3ed90) at /home/vagrant/php-src/main/main.c:2537
#33 0x0000000000c30a9f in do_cli (argc=2, argv=0x2dd6490) at /home/vagrant/php-src/sapi/cli/php_cli.c:993
#34 0x0000000000c31d1d in main (argc=2, argv=0x2dd6490) at /home/vagrant/php-src/sapi/cli/php_cli.c:1381
```

**Comment by SirSnyder** [comment]
@krakjoe @tpunt any idea how to fix this issue?

**Comment by tpunt** [comment]
@SirSnyder I have taken a brief look into this, but I haven't gotten very far regarding your problem. When simplifying your script, I hit two further segfaults, both on shutdown of the engine (one seems related to traits and the other one is in the GC root buffer). I may be able to look more into this today/tomorrow.

**Comment by SirSnyder** [comment]
@tpunt how did you simplified the script? My example script segfaults for me at runtime, not at shutdown. After 8 calls of ```$this->objStack[] = $objInstance;``` in ManagedInstanceProvider::manage() a zend_hash_do_resize() crashes with 
> zend_mm_heap corrupted
> Segmentation fault (core dumped)

If I remove ```private $intObjectHash;``` from ```trait ClassInfo```, remove ```use ClassInfo;``` from ```class ThreadedMap``` or replace ```new ThreadedMap``` by ```new Threaded()``` in line 48 everything is fine but such a fix is no proper solution. I've no clue where to start debugging in pthreads source.

**Comment by tpunt** [comment]
@SirSnyder 
> how did you simplified the script?

I just played around with it, seeing what I could delete/change and still get it to segfault in the same way.

> My example script segfaults for me at runtime, not at shutdown.

Yes, I know your script does not segfault during the shutdown sequence of the Zend Engine. Those are simply the segfaults I received further to simplifying your reproducing code.

The logic in your code is also broken in places. For example, take the logic in your `ManagedInstanceProvider::get` method:
```php
public function get()
{
    return $this->objStack->synchronized(function ()
    {
        $objInstance = null;

        while($objInstance === null)
        {
            if($this->objStack->count() > $this->intLastIndex)
            {
                // get instance
                $objInstance = $this->objStack[$this->intLastIndex];

                ++$this->intLastIndex;
                continue;
            }
            $this->objStack->wait(0);
        }
        return $objInstance;
    });
}
```

The `while` loop condition will *always* be `true` because you are always fetching a non-existent index from the `$this->objStack` object. This is due to the trait in the `ThreadedMap` class, which offsets the property count by 1. So what you are pretty much doing is the following cycle:
 - Does `$this->intLastIndex` (0) equal the property count of `$this->objStack` (1)? (start while loop)
   - Nope. Check (non-existent) property at offset `$this->intLastIndex` (0) exists?
   - Nope. Increment `$this->intLastIndex` (1)
   - Does `$this->intLastIndex` (1) equal the property count of `$this->objStack` (1)?
   - Yep. Wait for a new object to be added to `$this->objStack`
   - Add a new object at `$this->intLastIndex` - 1, since the original index (0) did not exist, and you then proceeded to increment `$this->intLastIndex` anyway.
 - [LOOPED LOGIC] Does `$this->intLastIndex` (1) equal the property count of `$this->objStack` (2)? (continue while loop)
   - Nope...

Thus, you have an infinite loop there. Performing a `var_dump` on the `$this->intLastIndex` from inside `ManagedInstanceProvider::manage` and on `$objInstance` from inside `ManagedInstanceProvider::get` further verifies this:
```
NULL
int(1)
NULL
int(2)
NULL
int(3)
NULL
int(4)
NULL
int(5)
NULL
int(6)
NULL
int(7)
NULL
int(8)
NULL
int(9)
```

There are also a few other problems (such as invoking `wait` in unconditioned blocks, which may lead to spurious wakeup calls). But anyway, back to the problem at hand (the segfault).

I've shortened your reproducing script to around 1/7 of the size to the following:
```php
class Test extends Threaded
{
    private $prop;
}

$objTestThread = new class extends Thread {
    public function __construct()
    {
        $this->test = new Test();
    }

    public function run()
    {
        $this[] = $this->test;
        while (1);
    }
};

$objTestThread->start();

while (true) {
    if (isset($objTestThread[0])) {
        $objTestThread[0][] = new Threaded();
    }
}
```

> After 8 calls of $this->objStack[] = $objInstance; in ManagedInstanceProvider::manage() a zend_hash_do_resize() crashes with

The problem occurs on the 9th insertion because PHP's hash tables have an initial size of 8 (properties are stored in a PHP hash table). When an attempt is made to insert a 9th element, it requires the hash table to be enlarged. This causes a reallocation, which is where the segfault occurs.

It appears that it is only triggerable if:
 - A new `Threaded` object with at least one default property is created in the main thread
 - This new `Threaded` object is assigned to a property of the `Thread` inside of the new thread
 - The newly assigned property (the `Threaded` object) from inside the thread is referenced from the main thread, where it has its property table enlarged from the main thread by assigned new `Threaded` objects.

Maybe it a simpler reproducing scenario can be found still. I think the problem comes down to reallocating the hashtable in a different thread to where it was originally allocated. Unfortunately, I haven't really had the time to look deeper into this yet.

> I've no clue where to start debugging in pthreads source.

I start by simplifying the reproducing *as much as possible*. From there, I piece together the important parts of the reproducing script, and then I follow the execution through pthreads by setting breakpoints at various functions in its source code.

**Comment by SirSnyder** [comment]
@tpunt 
Initialization of the hashmap happens in store.c line 319
```
if (Z_TYPE_P(write) == IS_OBJECT) {
		/* when we copy out in another context, we want properties table
			to be initialized */
		rebuild_object_properties(Z_OBJ_P(write));
}
```
Your example script runs fine without `rebuild_object_properties` but I guess it's not that easy. Dropping the line let the test verify-overload.phpt fail. Interesting fact by the way. I followed `rebuild_object_properties` and in case of default properties `rebuild_object_properties` calls `zend_hash_real_init_ex` which inits the default 8 slots. Maybe this could be a starting point to fix the issue but I doesn't know how, I'm not familiar with the php internals as well as hashtables. On the other hand the approach could be totally wrong and it's an issue with something else, handlers for example. The issue came up with php 7.1. So what has been changed between 7.0 and 7.1 at the hashtable stuff?
